custom_image_utils/shell_script_generator.py

```diff
@@ -35,32 +35,51 @@
   local -r cmd="$*"
 
   for ((i = 0; i < 3; i++)); do
-    if eval "$cmd"; then return 0 ; fi
+    time eval "$cmd" > "/tmp/{run_id}/install.log" 2>&1 && retval=$? || {{ retval=$? ; cat "/tmp/{run_id}/install.log" ; }}
+    if [[ $retval == 0 ]] ; then return 0 ; fi
     sleep 5
   done
   return 1
 )
 
+function gsutil() {{ ${{gsutil_cmd}} $* ; }}
+
+function version_ge() ( set +x ;  [ "$1" = "$(echo -e "$1\n$2" | sort -V | tail -n1)" ] ; )
+function version_gt() ( set +x ;  [ "$1" = "$2" ] && return 1 || version_ge $1 $2 ; )
+function version_le() ( set +x ;  [ "$1" = "$(echo -e "$1\n$2" | sort -V | head -n1)" ] ; )
+function version_lt() ( set +x ;  [ "$1" = "$2" ] && return 1 || version_le $1 $2 ; )
+
+function prepare() {{
+  # With the 402.0.0 release of gcloud sdk, `gcloud storage` can be
+  # used as a more performant replacement for `gsutil`
+  gsutil_cmd="gcloud storage"
+  gcloud_sdk_version="$(gcloud --version | awk -F'SDK ' '/Google Cloud SDK/ {{print $2}}')"
+  if version_lt "${{gcloud_sdk_version}}" "402.0.0" ; then
+    gsutil_cmd="$(which gsutil) -o GSUtil:check_hashes=never"
+  fi
+}}
+
 function exit_handler() {{
   echo 'Cleaning up before exiting.'
 
-  if [[ -f /tmp/{run_id}/vm_created ]]; then
+  if [[ -f /tmp/{run_id}/vm_created ]]; then ( set +e
     echo 'Deleting VM instance.'
-    execute_with_retries gcloud compute instances delete {image_name}-install \
-        --project={project_id} --zone={zone} -q
-  elif [[ -f /tmp/{run_id}/disk_created ]]; then
+    execute_with_retries \
+      gcloud compute instances delete {image_name}-install --project={project_id} --zone={zone} -q
+  ) elif [[ -f /tmp/{run_id}/disk_created ]]; then
     echo 'Deleting disk.'
-    execute_with_retries gcloud compute ${{base_obj_type}} delete {image_name}-install --project={project_id} --zone={zone} -q
+    execute_with_retries \
+      gcloud compute ${{base_obj_type}} delete {image_name}-install --project={project_id} --zone={zone} -q
   fi
 
   echo 'Uploading local logs to GCS bucket.'
-  gsutil -m rsync -r {log_dir}/ {gcs_log_dir}/
+  gsutil rsync -r {log_dir}/ {gcs_log_dir}/
```

Comment:
It looks like this has dropped the `-m` flag if the invocation is using legacy `gsutil` instead of `gcloud storage`. This might impact performance.

Reply:
`gcloud storage` considers the `-m` implied and will fail if it is passed

Reply:
Thanks for talking with me about this.  I understand better now what you're concerned about.  I'll look for other instances in the other files with `gcloud` commands included.